### ISSUE TITLE
builtins.readDir does not honor --store

### ISSUE THREAD

**Issue by bluescreen303** [NixOS/nix#3956]
The following snippet works fine using normal `nix-build`, but fails when using `nix-build --store /mnt`. 

```
{pkgs ? import <nixpkgs> {}, lib ? pkgs.lib}:
with lib;

let foo = pkgs.runCommand "dir" {} ''
              mkdir -p $out/foo
              mkdir -p $out/bar
          '';

in pkgs.writeText "dir.json" (builtins.toJSON (builtins.readDir foo))
```

**Error**

```
error: opening directory '/nix/store/8g66y9n449pq8n46xrqc178nbqhjnrbf-dir': No such file or directory
```

**Expected behavior**

```
building '/nix/store/7l4a0ajxsx1pvazc5l3238db9p7cq0gp-dir.drv'...
these derivations will be built:
  /nix/store/61hz4ikmhghmbx8cx8savs8rxx4ddlj9-dir.json.drv
building '/nix/store/61hz4ikmhghmbx8cx8savs8rxx4ddlj9-dir.json.drv'...
/nix/store/2dpvnqprj8yym5rpm154bj9fh3vd60wz-dir.json
```

**Warning**
When running without `--store` once, the directory gets created in `/nix/store`, which makes any following tries **with** `--store` succeed. So watch out when trying to reproduce and always delete the "dir" derivation from the store.

**`nix-env --version` output**

nix-env (Nix) 2.3.7

**Comment by Ericson2314** [comment]
Hmm, this is an interesting problem. `--store`, by default, create a store which, while stored elsewhere, is still mounted (in a chroot) at `/nix/store`. The problem is that building, and not eval, occurs in that chroot. `builtins.readFile` is also meant to pick up various impurities, so it is normal for it to not use a chroot-root-relative path.

Internal to Nix, there is a function "store.toRealPath" which will modify a path to include the path to the dir mounted as root, but I am not sure how it could get used here automatically. Maybe there is a builtin for it?

**Comment by bluescreen303** [comment]
I searched for it, but didn't find it. `builtins.storeDir` is always "/nix/store" and builtins.storePath also does not seem to return a real path.

Yes, I understand `readFile` and `readDir` are meant to pick up impurities, that's what I use it for. But since /nix/store has been shadowed when running with `--store` I don't think there's a case for wanting to read the original during eval. But even when such a case does exist, passing it a derivation should be a clear indication a path from the `--store` version is requested.

**Comment by Ericson2314** [comment]
> But even when such a case does exist, passing it a derivation should be a clear indication a path from the --store version is requested.

But that's the thing, `builtins.readFile` consumes an arbitrary string or path. Neither `builtins.readFile` or `builtins.derivationStrict` has enough context to automatically do what you want, I'm afraid.

**Comment by bluescreen303** [comment]
Which is why I reported this one :)

I understand readFile/readDir take a path and not a derivation. But there could be some logic that recognizes the requested path is in `/nix/store` which has been mapped over by `--store` and translate to the real path.

I tried to manually do this, but nix doesn't let me append a path with context onto another path.
Luckily, it does work as a string and readFile/readDir do accept strings, but I feel this is hacky.

- I need to pass in the `--store` path to the evaluator
- I have no way of modifying paths into the actual paths. Yes, string workaround works, but what if some function really wants a path?

**Comment by Ericson2314** [comment]
@bluescreen303 I don't think we should automatically convert any part which happens to make sense in the chroot, because it's possible that it also makes sense outside the chroot. Yes, that possibility is unlikely, but I rather not do spooky-magic designs justified by probability.

Adding the `builtins.toRealPath` that I hoped might exist I think is a good simple solution, whose boring and manual usage comforts me.

**Comment by bluescreen303** [comment]
I agree. Sounds like a simple and safe solution :)

**Comment by Ericson2314** [comment]
I wonder if lazy tree's virtualization of the file system can offer a nicer solution than `builtins.toRealPath` CC @roberth.
